kernel/sem.c

```diff
@@ -98,18 +113,165 @@ int z_vrfy_k_sem_init(struct k_sem *sem, unsigned int initial_count,
 static inline void handle_poll_events(struct k_sem *sem)
 {
 #ifdef CONFIG_POLL
+#ifdef CONFIG_SEM_IMPL_ATOMIC
+	k_spinlock_key_t key = k_spin_lock(&poll_lock);
+
+#endif
 	z_handle_obj_poll_events(&sem->poll_events, K_POLL_STATE_SEM_AVAILABLE);
+#ifdef CONFIG_SEM_IMPL_ATOMIC
+	k_spin_unlock(&poll_lock, key);
+#endif
 #else
 	ARG_UNUSED(sem);
 #endif
 }
 
-void z_impl_k_sem_give(struct k_sem *sem)
+#ifdef CONFIG_SEM_IMPL_ATOMIC
```

Comment:
Style: rather than the giant #if blocks, it might be clearer if you had (static inline) implementations for k_sem_give_atomic, k_sem_give_1lock, k_sem_give_lock, etc... and then select between them in a much tighter selection at the bottom of the file where you define the real API in terms of the others.

Reply:
Sure, I can do that. Only reason I didn't previously is because of the public `struct k_sem` data structure part... any suggestions on that front? When I try to use a typedef checkpatch complains.

Reply:
Ok, so I tried this, and ran into a few different issues:

1. `k_sem` is publicly exposed, so any e.g. typedef or #define would also have to be publicly exposed.
2. The different variants of `k_sem` are _almost_ identical, and end up with a lot of duplication of the data structure elements (for e.g. polling).
3. `*_one_lock` and `*_glob_lock` are pretty much identical, but end up duplicated here.
4. There's no good way to interface between the toplevel API taking `k_sem` and the implementations taking e.g. k_sem_atomic.

A quick example / illustration of the problems I'm running into here, for just the 2 not 3 implementations:

```c
/* kernel.h */

/**
 * @cond INTERNAL_HIDDEN
 */
struct k_sem_single_lock
{
	_wait_q_t wait_q;

	unsigned int count;
	unsigned int limit;

	_POLL_EVENT;

	_OBJECT_TRACING_NEXT_PTR(k_sem)
	_OBJECT_TRACING_LINKED_FLAG
};


/**
 * @cond INTERNAL_HIDDEN
 */
struct k_sem_atomic
{
	_wait_q_t wait_q;

	atomic_t count;
	atomic_t limit;

	_POLL_EVENT;

	_OBJECT_TRACING_NEXT_PTR(k_sem)
	_OBJECT_TRACING_LINKED_FLAG
};

#if defined(CONFIG_SEM_IMPL_ATOMIC)
#define k_sem k_sem_atomic
#else
#define k_sem k_sem_single_lock
#endif
```

```c
/* ksem.c */

static inline void sem_give_lock(struct k_sem_lock *sem)
{
    [...]
}

static inline void k_sem_give_atomic(struct k_sem_atomic *sem)
{
    [...]
}

void z_impl_k_sem_give(struct k_sem *sem)
{
	sys_trace_semaphore_give(sem);

#if defined(CONFIG_SEM_IMPL_ATOMIC)
	sem_give_atomic(sem);
#else
	sem_give_lock(sem);
#endif

	sys_trace_end_call(SYS_TRACE_ID_SEMA_GIVE);
}
```
...only this then ends up with the _next_ issue, namely that now the compiler errors out complaining that sem_give_atomic (or sem_give lock, depending) isn't used.

Any suggestions here?